### DISCUSSION TITLE
What about further extending JsFuture to adapt more callback APIs?

### DISCUSSION THREAD

**Discussion by LittleSaya** [wasm-bindgen/wasm-bindgen#4261] [Ideas]
I found this type can be easily used on different callback APIs with just a little bit modification, maybe this could become a more general method to wrap callback APIs into Futures?

examples:
- [FileSystemDirectoryEntry](https://github.com/LittleSaya/wasm-zip-stream/blob/main/src/js_futures/FileSystemDirectoryEntry_getFile_future_mod.rs)
- [FileSystemFileEntry](https://github.com/LittleSaya/wasm-zip-stream/blob/main/src/js_futures/FileSystemFileEntry_file_future_mod.rs)
- [FileSystemDirectoryReader](https://github.com/LittleSaya/wasm-zip-stream/blob/main/src/js_futures/FileSystemDirectoryReader_readEntries_future_mod.rs)

**Comment by daxpedda** [comment]
I'm a bit lost on what you are proposing here. I would appreciate if you explain it from first principles to me.

**Reply by LittleSaya** [reply]
I think this will make using callback APIs easier (write fewer codes)

For example, for different kinds of callback APIs, we have different helper functions:

| API | helper function |
|--------|--------|
| 1 callback | `fn helper1(obj: &JsValue, method: &str, args: js_sys::Array) -> Future<Output = JsValue>` |
| 2 callback | `fn helper2(obj: &JsValue, method: &str, args: js_sys::Array) -> Future<Output = (JsValue, JsValue)>` |
| 3 callback | `fn helper3(obj: &JsValue, method: &str, args: js_sys::Array) -> Future<Output = (JsValue, JsValue, JsValue)>` |

In each helper, we can retrieve `method` from `obj` and cast it into a `Function`, then we create an `Array` of `Closure`s according to the number of callbacks and concatenate the `Closure` array with `args`, then we invoke the `Function` with this new concatenated array, finally cleanup `Closure`s and fulfill the `Future`.

I think the problem would be the position of callback arguments, I am assuming that callback parameters are always the last arguments, I do not know if there is any API that puts callbacks before other arguments.

**Reply by daxpedda** [reply]
I see. However this isn't in scope for `wasm-bindgen-futures`. This project is only here to provide low-level bindings, anything beyond that should be done in higher-level crates.

What is in scope however is https://github.com/rustwasm/wasm-bindgen/issues/4116, which would allow [`Promise::then()`](https://docs.rs/js-sys/0.3.72/js_sys/struct.Promise.html#method.then) to also convey these types in the returned `Promise`.

This should address your issue as well, right?

**Reply by LittleSaya** [reply]
Emm, typing is not my primary concern, my real hope is erasing as much boilerplate codes as possible, but the `Promise` type you mentioned just inspired me, that, actually it is much easier to wrap those callback APIs into Promise in javascript side, and just use `await_promise` and use `extern "C"` to import wrapped APIs in rust side, this could eliminate a lot of glue codes and achieve the goal of 'write fewer codes' as well.
